### DISCUSSION TITLE
[Seção 4.21] Dúvidas sobre Modules - filename -  variável var.files -

### DISCUSSION THREAD

**Discussion by fernandomullerjr** [chgasparoto/curso-aws-com-terraform#108] [Q&A]
Oi Cleber, boa noite!

Estou com 2 dúvidas sobre a aula 21, que trata sobre o assunto Modules, por volta dos 20min de aula.

- Dúvida 1

Sobre o arquivo outputs da pasta s3_module, fiquei na dúvida como o Data Source "module.objects" é utilizado na expressão "for".

- Arquivo de referência:
`/s3_module/outputs.tf`

-Trecho do código:
```
output "files" {
  value = [for filename, data in module.objects : filename]
}
```

No caso o primeiro nome "filename" é só um nome para os valores que são trazidos pelo Data Source, certo? Só que aquele segundo "filename" ao final da
expressão "for", o que ele faz? Não entendi aonde exatamente diz para a expressão trazer só o filename do Data Source, visto que olhando a documentação da Terraform sobre o Data Source "aws_s3_bucket_object"
Não há alguma indicação deste "filename"
Consegue me explicar como é obtido o valor deste output files, por favor?

- Dúvida 2

No arquivo main.tf do módulo s3_module existem alguns campos onde é usada a variável var.files
Verificando o arquivo variables.tf da pasta s3_module, essa variável var.files é iniciada com um valor default vazio.
Não consegui entender da onde o Terraform obtém o valor para a variável var.files que é usado no for_each e na no campo src do módulo objects.
Devido isto, não consegui entender como o for_each vai funcionar, porque em nenhum momento informei um valor para o var.files ou indiquei algum lugar para ele obter um valor.
Consegue me ajudar a entender melhor sobre como esta variável está sendo usada, por favor?

- Arquivo de referência:
`/s3_module/main.tf`

- Trecho de código:

```
module "objects" {
  source = "./s3_object"

  for_each = var.files != "" ? fileset(var.files, "**") : []

  bucket = aws_s3_bucket.this.bucket
  key    = "${var.key_prefix}/${each.value}"
  src    = "${var.files}/${each.value}"
}
```

Obrigado desde já!

**Comment by chgasparoto** [comment]
@fernandomullerjr vamos lá:

**Dúvida 1**
O módulo de `objects` retorna um `map` pois ele tem um `[for_each](https://github.com/chgasparoto/curso-aws-com-terraform/blob/master/02-terraform-intermediario/05-modules/s3_module/main.tf#L29)` quando ele é chamado. A chave desse `map` é o `nome do arquivo` ou o `filename`.

Exemplo:

```hcl
module.objects = {
  "meu_site/index.html" = {
   // aqui vai os outputs do módulo
  },
  "meu_site/error.html" = {
   // aqui vai os outputs do módulo
  },
}
```

Aqui você pode ver todos os [outputs](https://github.com/chgasparoto/curso-aws-com-terraform/blob/master/02-terraform-intermediario/05-modules/s3_module/s3_object/outputs.tf).

Se você olhar na [documentação](https://www.terraform.io/language/expressions/for#input-types), a sintaxe para iterar sobre um map ou objeto é usando `for chave, valor in map...` ou `for parametro1, parametro2 in map...`

Ali, `chave` e `valor` pode ser o que você quiser. Eu poderia muito bem colocar `arquivo` ou `f`. Esses dois paramêtros são variáveis e você seleciona o nome de cada um.


**Dúvida 2**
A variável `var.files` recebe o caminho ou path de onde estão os arquivos que você quer subir para o S3.

No trecho de código abaixo eu estou utilizando uma função do terraform para ler um diretório e retornar uma lista de arquivos encontrados.

`for_each = var.files != "" ? fileset(var.files, "**") : []`

É a partir do retorno desta função que o `for_each` interage pois a função [fileset](https://www.terraform.io/language/functions/fileset) retorna uma lista de arquivos com o caminho absoluto ao valor dado à variável `var.files`.

**Comment by fernandomullerjr** [comment]
Oi Cleber, boa noite.

Obrigado pelo retorno.

Ainda fiquei com algumas dúvidas sobre a segunda dúvida.
Seguem os detalhes:

**Dúvida 2**

Sobre a variável files, você menciona que:

> É a partir do retorno desta função que o `for_each`  interage pois a função [fileset](https://www.terraform.io/language/functions/fileset) retorna uma lista de arquivos com o caminho absoluto ao valor dado à variável var.files.

Eu não consegui entender o final sobre o valor dado a variável var.files, porque a principio não é dado nenhum valor a ela.

Único momento que vi algo com nome "files" receber algum valor, foi no atributo do module "website":
`files = "${path.root}/website"`

Mas não consegui ver como ele popula o `var.files`, para que o `fileset` consiga visualizar que o `var.files` não esteja vazio e seguir com a lógica do na sequência.

Consegue me ajudar nessa parte novamente, por favor?

**Reply by chgasparoto** [reply]
@fernandomullerjr 

A função `fileset` recebe um caminho, que está sendo passado [aqui](https://github.com/chgasparoto/curso-aws-com-terraform/blob/master/02-terraform-intermediario/05-modules/main.tf#L39), e retorna um array com os arquivos encontrados conforme o `pattern` (`fileset(var.files, "**")` -> pattern `**` quer dizer qualquer arquivo, incluindo subdiretórios) que é passado. 

No caso do módulo, eu estou passando para ele o caminho onde estão os arquivos do website (`files = "${path.root}/website"`) e a função `fileset` vai retornar um array com dois arquivos:

```hcl
[
  "website/index.html",
  "website/error.html"
]
```

Dentro do [s3_module](https://github.com/chgasparoto/curso-aws-com-terraform/blob/master/02-terraform-intermediario/05-modules/s3_module/main.tf#L29) ele vai iterar sobre esses dois arquivos e criar dois recursos do tipo `bucket object` com os respectivos arquivos dentro do diretório `website` (`index.html` e `error.html`).

Se ainda não ficou claro me avise. Abraço.

**Comment by fernandomullerjr** [comment]
Oi Cleber, boa tarde.

Primeiro, muito obrigado, teus esclarecimentos ajudaram demais, esse segundo esclareceu 99% das dúvidas.
Mas como você disse para avisar se não ficar claro algo, só tem 1 detalhe que fiquei na dúvida ainda.

Na estrutura do projeto, vem a raíz e depois as pastas `s3_module ` e `s3_object`.

├── main.tf
├── outputs.tf
├── s3_module
│   ├── main.tf
│   ├── outputs.tf
│   ├── README.MD
│   ├── s3_object
│   │   ├── main.tf
│   │   ├── outputs.tf
│   │   └── variables.tf
│   └── variables.tf
├── terraform.tfstate
├── terraform.tfstate.backup
└── website
    ├── error.html
    └── index.html

Só não entendi como a `var.files` tem o valor do campo files lá da raíz, visto que o `var.files` é usado dentro do `main.tf` da pasta `s3_module`.
Seria pelo fato do `main.tf` chamar o módulo da `s3_module ` através do `source = "./s3_module"` ?
Porque no meu entendimento, o `var.files` no `main.tf` do s3_module seria o valor da variável dentro do arquivo `variables.tf` no mesmo escopo/pasta do `main.tf` do `s3_module`.
No mais, obrigado mais uma vez pela tua ajuda, só restou esta pequena dúvida mesmo, porque os conceitos dos módulos deram uma confundida na cabeça mesmo.
Abraço.

**Reply by jeffersonfragoso** [reply]
@fernandomullerjr 
Você já deve ter entendido o funcionamento mas farei alguns comentários para quem chegar depois porque tive uma dúvida parecida logo no início da aula, "como a variável 'website' do s3_module foi setada"? 

Esse é o funcionamento padrão dos [modulos do terraform](https://developer.hashicorp.com/terraform/language/modules/develop) junto com o  conceito de [ input variables](https://developer.hashicorp.com/terraform/language/values/variables). 

Dentro de um bloco 'module' é possível setar os valores das variáveis do modulo referenciado em 'source', bastando definir chaves com os mesmo nomes das variáveis do modulo referenciado. Como a documentação diz é como se fosse uma  passagem de argumentos de função. 
Acho isso bem confuso, preferiria que existisse algo mais explícito, talvez 'source.var.minha_variavel = '.

Sobre esse ponto:

> o var.files no main.tf do s3_module seria o valor da variável dentro do arquivo variables.tf no mesmo escopo/pasta do main.tf do s3_module.

Até agora no curso tinhamos visto que era possível setar os valores das variáveis por linha de comando ou pelo arquivo .tfvars, pois bem, uma terceira forma é como comentado acima utilizado a sintaxe do bloco 'module'.
No final das contas var.files no main.tf do s3_module teve seu valor setado em runtime [nesse](https://github.com/chgasparoto/curso-aws-com-terraform/blob/master/02-terraform-intermediario/05-modules/main.tf#L39) momento. Possuindo o caminho completo para os arquivos o terraform por debaixo dos panos deve executar alguma api do go para conseguir lidar com arquivos, mesmo em outra pasta.
